### PR TITLE
Remove ThemeProps and add new theme state logic

### PR DESCRIPTION
**Pull request by vovakulikov**
## Before

Before this PR we had theme state on the top level in the Layout.tsx, we had a hook that provided an object with public API for theme (to read or to set a new value), there were at least two problems 
- Complex implementation of useThemeProps (rxjs to listen browser theme, mixed with logic around setting theme CSS classes)
- Prop drilling, after we created these theme API on the top level, it was distributed via props

I think I don't need to dive deep in why prop drilling is a bad practice, but still, with prop drilling and nested components, we were updating the whole react tree when theme was changed, also passing these public Theme props required extending your component props with special type `ThemeProps` and this becomes a mess quickly if you extend something else (like one component props extended other component props)

## Now

In this PR, we use more standard ways for global state distirubiotion - Context. In the new `useTheme` hook, we don't do any effects with CSS classes (consumer on the top level should handle it), so it's safe to call useTheme on any UI level. Privosly it wasn't safe, and because of this, we had this problem https://github.com/sourcegraph/sourcegraph/issues/47501

## Further steps

Ideally we shouldn't have any JS logic about theme besides CSS classes set on the document element. And all colors should work through CSS variables *custom-properties, but at the moment, we have at least two major places where we have to have JS check about theme. 
- Brand logo (since we use svg as image, in img src attribute), we should change URL based on theme (in dark theme brand logo has difference color value). I'm wondering that we probably can change it and use svg directly without img tag)
- Monaco editor doesn't support colors through CSS variables and we change set of colors in js runtime based on theme value, I don't know maybe with CodeMirror we could use just CSS variables and remove this logic 

## Test plan
- Check web app with different theme values
- Check jetbrains plugin 
- Check vscode webview extension 
- Check browser extension UI 
- Check code host integration tooltip UI 

## App preview:

- [Web](https://sg-web-vk-improve-theme-state.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.00% (-0.11 kb) | -0.01% (-0.99 kb) | -0.01% (-0.88 kb) | 0.13% (+1) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/vk/improve-theme-state/compare-report.html;tab=live_object) for a full comparison between the commits 5c57ba8 and 59c3fa5 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/browser/src/shared/code-hosts/shared/codeHost.tsx [hunk 1]
```diff
@@ -525,13 +520,13 @@ function initCodeIntelligence({
...
                     .subscribe()
             )
-            if (codeHost.isLightTheme) {
-                this.subscription.add(
-                    codeHost.isLightTheme.subscribe(isLightTheme => {
-                        this.setState({ isLightTheme })
-                    })
-                )
-            }
+            // if (codeHost.isLightTheme) {
```
Comment (vovakulikov) [inline client/browser/src/shared/code-hosts/shared/codeHost.tsx:523]: We probably don't need this anymore

**Comment by philipp-spiess** [comment]
I’m going to do a more detailed review tomorrow but I want to comment and say how excited I am about this change. I have wished for a `useTheme` hook before: This is a very welcome change. 🙂

File: client/web/src/enterprise/embed/EmbeddedWebApp.tsx [hunk 2]
```diff
@@ -35,21 +35,23 @@ const EmbeddedNotebookPage = lazyComponent(
...
-    const isLightTheme = enhancedThemePreference === ThemePreference.Light
+    const { theme, setThemeSetting } = useTheme()
+
+    useLayoutEffect(() => {
+        const isLightTheme = theme === Theme.Light
+
+        document.documentElement.classList.add('theme')
+        document.documentElement.classList.toggle('theme-light', isLightTheme)
+        document.documentElement.classList.toggle('theme-dark', !isLightTheme)
+    }, [theme])
```
Comment (philipp-spiess) [inline client/web/src/enterprise/embed/EmbeddedWebApp.tsx:46]: Should we extract this into a hook as well? Seems like we need it in a few places

**Review by philipp-spiess — APPROVED** [review]
Oprah Crazy

File: client/shared/src/theme.ts [hunk 3]
```diff
@@ -1,35 +1,134 @@
...
+ * of components)
  */
-export const observeSystemIsLightTheme = (
-    window_: Pick<Window, 'matchMedia'> = window
-): { observable: Observable<boolean>; initialValue: boolean } => {
-    const mediaList = window_.matchMedia('(prefers-color-scheme: dark)')
+export function useTheme(): useThemeApi {
+    const { themeSetting: contextThemeSetting, onThemeSettingChanging } = useContext(ThemeContext)
+
+    const systemTheme = useSyncExternalStore(subscribeToSystemTheme, getSystemThemeSnapshot)
```
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/shared/src/theme.ts:64]: Chef's kiss!

File: client/web/src/site-admin/SiteAdminPingsPage.tsx [hunk 4]
```diff
@@ -15,18 +15,19 @@ import {
...
 import { eventLogger } from '../tracking/eventLogger'
 
-interface Props extends ThemeProps {}
+interface Props {}
 
 /**
  * A page displaying information about telemetry pings for the site.
  */
-export const SiteAdminPingsPage: React.FunctionComponent<React.PropsWithChildren<Props>> = ({ isLightTheme }) => {
+export const SiteAdminPingsPage: React.FunctionComponent<React.PropsWithChildren<Props>> = () => {
```
Comment (valerybugakov) [inline client/web/src/site-admin/SiteAdminPingsPage.tsx:29]: After removing most of the prop drilling, I imagine this is how most of the components will look. No props!

File: client/web/src/nav/UserNavItem.test.tsx [hunk 5]
```diff
@@ -58,10 +56,7 @@ describe('UserNavItem', () => {
                 <MemoryRouter>
                     <CompatRouter>
                         <UserNavItem
-                            isLightTheme={true}
-                            onThemePreferenceChange={() => undefined}
                             showKeyboardShortcutsHelp={() => undefined}
-                            themePreference={ThemePreference.Light}
```
Comment (valerybugakov) [inline client/web/src/nav/UserNavItem.test.tsx:64]: 🌪️

**Review by valerybugakov — APPROVED** [review]
This is huge, @vovakulikov! Thank you for looking into it 🔥

**Comment by philipp-spiess** [comment]
> BUT i think having separate stores-hooks isn't a best idea here because you have to know about them to use it. Imagine if we shared it with one hook/context where we can fine all that we may need (and we had suggestions there)

Hmmm I think I personally would prefer individual hooks for all of these because they don't really share logic and some of these (e.g. theme) might be updated where as others (e.g. telemetry stuff) are never changed which could cause unnecessary re-renderings when we have to subscribe to all context values.

But yeah everything is better than the current prop-drilling 😅

**Comment by vovakulikov** [comment]
@valerybugakov, it seems like there is a problem with vscode extension. I couldn't manage to run it locally due to problem with depencencie resolver, and I noticed that there are two deps that are not resolved 
```
 "@vscode/codicons": "^0.0.29",
 "@vscode/webview-ui-toolkit": "^0.9.0",
```
but then `"@microsoft/fast-web-utilities": "5.1.0"` also is not found (I tried to add it to deps, but this didn't help), If you got a minute can you take a look (apparently nobody uses this locally anymore I think, this is why we didn't catch thing during pnpm migration)

**Comment by vovakulikov** [comment]
Ok I think I got deps problem, so I manged to build vscode extension
we have to add two dependencies explicitly on the top level 
```
  "@microsoft/fast-web-utilities": "5.1.0",
  "@microsoft/fast-element": "1.7.0",
```
I think this is the ghost deps package problem, and this used to work before only because of how yarn and node deps resolution works (and how it doesn't work now with pnpm resolvers)

But then when I try to debug this locally I see blank screen (no runtime errors found in the terminal), does anyone know here what it may be? 

https://user-images.githubusercontent.com/18492575/220216188-4131db41-c255-49fa-b992-a17692b18c5d.mov

**Comment by vovakulikov** [comment]
A few thoughts on jetbrains plugin 
- It seems that esbuild (which is used by default for pnpm watch is failing with Markdown CSS (and this script just doesn't work)
- pnpm standalone produces localhost with non working version of search UI (see screenshot below)
Screenshot 2023-02-20 at 22 18 51
- And unfortunately at the end I couldn't run the plugin in sandboxes because java 11 isn't compatible with java version of IDE tookit. 

@philipp-spiess I remember you used to work with JB plugins, let me know if you have a minute to guide me through the standard process (maybe there is some catch with version that I'm missing)

**Comment by vovakulikov** [comment]
I managed to test code host integrations and browser extensions though, so it seems that after vscode and jb plugins we're good to go here

**Comment by philipp-spiess** [comment]
@vovakulikov 

Regarding VS Code: No I don't remember seeing the blank screen before 😞  I also don't recall having had an issue with dependencies, @taras-yemets and me have made a change in vscode not too long ago!

Regarding JetBrains: I can't get it to run myself either. It's crashing with this issue now 🤔 

Screenshot 2023-02-21 at 14 52 23

This seems unrelated to your change though. I wonder what goes on. I'll need more time to look into this I’m afraid 😭

**Comment by philipp-spiess** [comment]
@vovakulikov It does look like the build script for JetBrains is broken, yeah. I don't see the dist folder being created

**Comment by vovakulikov** [comment]
Here is a crazy idea (maybe not so bad actually if nobody has any things against it). What if we just forgot about vscode and jetbrains plugins in this PR? I know that these two are not a priority at the moment (in fact, I haven't seen any big/or any changes in these plugins for a long time) 

So I propose to merge this PR and deal with possible regressions about theme in VSCode and JB plugin when we fix the vscode and jetbrains plugins). I would like to fix them, but at the moment, I don't have capacity even to dive deep there (maybe it's one liner fix, who knows). 

cc @valerybugakov @philipp-spiess

**Comment by philipp-spiess** [comment]
> Here is a crazy idea (maybe not so bad actually if nobody has any things against it). What if we just forget about vscode and jetbrains plugins in this PR? I know that these two are not a priority at the moment (in fact, I haven't seen any big/or any changes in these plugins for a long time)
>
> So I propose to merge this PR and deal with possible regressions about theme in VSCode and JB plugin when we fix the vscode and jetbrains plugins). I would like to fix them, but at the moment, I don't have capacity even to dive deep there (maybe it's one liner fix, who knows).

This sounds reasonable to me, it's very expensive to keep this branch open and we can't miss out on the benefits it brings. I would also add that a regression because of the themes is not very likely given that this is not a high risk change. WDYT @taras-yemets?

**Review by valerybugakov — APPROVED** [review]
> apparently, nobody uses this locally anymore, I think; this is why we didn't catch thing during pnpm migration

Eh, that's alarming. We should add this build to our CI pipeline! I created an issue for that:
- https://github.com/sourcegraph/sourcegraph/issues/48012

Agree with the proposal to merge this PR as is and create a follow-up issue to verify that the JetBrains extension works as expected later once we fix the build.

**Review by vdavid — APPROVED** [review]
Wow, very nice! I only read the desc and scrolled through the changes, but it's very promising to make our lives easier on the front end! Thanks Vova! ❤️

**Comment by vovakulikov** [comment]
Ok, I'm about to merge this because I already got two big conflicts with main, so if something will show up because of this PR let me know

**Pull request merged**